core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -128,15 +154,28 @@ public void clusterChanged(ClusterChangedEvent event) {
                     } else {
                         indexSettings = indexService.getIndexSettings();
                     }
-                    if (indicesService.canDeleteShardContent(shardId, indexSettings)) {
-                        deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case CAN_DELETE:
+                            deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                            break;
+                        case NO_FOLDER_FOUND:
+                        case SHARED_FILE_SYSTEM:
```

Comment:
I wonder if we should mix shared file systems into this. It makes the logic more complex and it doesn't really save us much because canDeleteShardContent is super fast in this case (it's a setting lookup). If remove it, we can call the cache map something like `folderNotFoundCache` which imo will be easier to follow (other names welcome :))